### ISSUE TITLE
Prompt provides no indication of changes for binary files

### ISSUE THREAD

**Issue by nedbat** [nojhan/liquidprompt#647]
Shell: 

Operating system: 

Liquidprompt version: 

### Steps to Reproduce

1. In a git repo, change a .png file
2. The prompt doesn't change to indicate there's a modified file, because git reports 0 lines changed for binary files.
3.

### Expected Behavior
As I recall, liquidprompt v1 had some kind of indication, though I forget what it was.

**Comment by Rycieos** [comment]
I haven't explicitly tested this, but it makes sense that it does this right now. I don't know about the claim that it didn't do this in v1, but I can check.

How would you expect the prompt to look in this case? Red branch name (showing changes), and a `(0,0)` showing no changed lines?

**Comment by nedbat** [comment]
I just tried it with the v1 code: it shows `master(+0/-0) ±` in the prompt.

**Comment by Rycieos** [comment]
Fair enough. I'll look into it.

So you just want it to work like how it used to?

**Comment by nedbat** [comment]
The old way was fine. I'm open to other possibilities.  I just need the prompt to clearly indicate if a file has been modified, even if it's a binary file.

**Comment by Rycieos** [comment]
Fixed. It ended up being a mistake I made in the abstraction layer over all VCSs, so it affected all of them.

Thanks for the report!

**Comment by nedbat** [comment]
Hmm, I think I am trying the latest code, and it doesn't seem fixed on my end?

**Comment by Rycieos** [comment]
It is working for me:
```
[mark:~/dev/liquidprompt/docs/theme] master+ ± cp default-short.png default-med.png
[mark:~/dev/liquidprompt/docs/theme] master(+0/-0)+ ±
```

The fix is only on the latest commit on `master`, 5885e57bd260ae388459a524793876923f1f0b2c.

**Comment by nedbat** [comment]
It works for me in bash. It doesn't work in zsh (or at least in my zsh). In fact, now text files don't show line counts in the prompt either.  I tried to trace through the code.  It might have something to do with the caching of the git shortstats?

**Comment by Rycieos** [comment]
Yikes, I can confirm that it's broken on Zsh. The lack of test coverage on Git functions is biting me again. Your findings are probably right, I'll dig into it.

**Comment by Rycieos** [comment]
Finally figured it out:
```zsh
function test_func {
  typeset foobar
  [[ -z ${foobar+x} ]]
  echo $?
}; test_func
```
returns `1`. This returns `0` on both Bash and ksh, so I cannot figure out why Zsh thinks that it is set.

I need to find a way to test for unset variables that works for all shells. `[[ -z ${var:+x} ]]` will work, but it also returns true for empty variables as well. That would probably work for this use case, but I would rather have unset work...

**Comment by Rycieos** [comment]
According to `man zshbuiltins`:
> Except as noted below for control flags that change the behavior, a parameter is created for each name that does not already refer to one.  When inside a function, a new parameter is created for every name (even those that already exist), and is unset again when the function completes.  See `Local Parameters' in zsh‐param(1).

The metioned section in `zsh‐param(1)` does not give any more hints.

So it would seem that when declaring a variable as local in Zsh, it sets it to the empty string.

As far as I can tell with the relevant POSIX standard: https://pubs.opengroup.org/onlinepubs/9699919799/utilities/V3_chap02.html#tag_18_06_02
There is no specification as to how these tests act when the variable is a local. And as for the `typeset` command:
> If the command name [is `typeset`], the results are unspecified.

So the solution is that these caches can only be treated as invalid if they are empty or unset, not just unset.

**Comment by Rycieos** [comment]
Fixed in 0ac2ad4a1b110536179ef5d10ca9fe48ea675421.

The worst part of this is that I already had a comment in the relevant tests describing this issue:
https://github.com/nojhan/liquidprompt/blob/0ac2ad4a1b110536179ef5d10ca9fe48ea675421/tests/test_shell.sh#L158-L159

I guess I forgot that in the past 3 months 🤦 

Thanks for the follow up report!

**Comment by nedbat** [comment]
Fix confirmed! Thanks :)
